assets/js/base/context/providers/cart-checkout/checkout-state/utils.ts

```diff
@@ -0,0 +1,63 @@
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { decodeEntities } from '@wordpress/html-entities';
+
+/**
+ * Internal dependencies
+ */
+import type { PaymentResultDataType, CheckoutResponse } from './types';
+
+/**
+ * Prepares the payment_result data from the server checkout endpoint response.
+ */
+export const getPaymentResultFromCheckoutResponse = (
+	response: CheckoutResponse
+): PaymentResultDataType => {
+	const paymentResult = {
+		message: '',
+		paymentStatus: '',
+		redirectUrl: '',
+		paymentDetails: {},
+	} as PaymentResultDataType;
+
+	// payment_result is present in successful responses.
+	if ( 'payment_result' in response ) {
+		paymentResult.paymentStatus = response.payment_result.payment_status;
+		paymentResult.redirectUrl = response.payment_result.redirect_url;
```

Comment:
Currently `response.payment_result.redirect_url` might not be present? So `CheckoutResponseSuccess` should probably have this defined as optional (see comment I left with that type) and thus there would need to be a check to make sure it's present here.

```suggestion
		paymentResult.redirectUrl = response.payment_result?.redirect_url;
```

Of course if you've discovered that `redirectUrl` is always in the response then this should be fine, but this  does significantly change behaviour because prior to this the `redirectUrl` was only used if it was non-truthy:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/a5add04cd927090c356136eaf2920b4d6f1827cc/assets/js/base/context/providers/cart-checkout/checkout-state/reducer.js#L117-L124

Won't the change in this PR mean that if a `paymentResult.redirectUrl` is set to an empty string then whatever is set as the redirectUrl in the state will get cleared given the reducer now explicitly checks for undefined?

Reply:
Added the truth check (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4200/commits/b2dc37e084bd2c7c69134fcb78d91da5e93177a3) and a followup todo to see what the implications of this are (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4200/commits/dd70b86396e04a98e26531bb397069bf7474b029)